### DISCUSSION TITLE
RTQ query result type based on query args

### DISCUSSION THREAD

**Discussion by bn-piar** [reduxjs/redux-toolkit#2701] [Q&A]
I want to have a query that would return different types of data based on args  
```javascript

const { data } = useGetDataQuery({ type: 'one' });
//      ^ type OneData
const { data } = useGetDataQuery({ type: 'two' });
//      ^ type TwoData
```

I tried 
```javascript

type Args = {
  type: 'one'
} | {
  type: 'two'
}
...
getData: builder.query<OneData | TwoData, Args>({...})
...
```
But obviously result type will be `OneData | TwoData` regardless the args passed. I need to somehow connect ResultType with QueryArgs but I don't see a way to do that. 

Use case: I have 5 pages that render tabular data. Queries for each one look like that: 
```javascript
// catalogue
query: (tab) => `catalogue/${tab}`
// reports 
query: (tab) => `reports/${tab}`
// so on
```
I wanted to make a hook that would accept a page and a tab and return corresponding data. So instead of 5 separate queries I'd have one
```javascript
query: ({ page, tab }) => `${page}/${tab}`
```

The problem with this approach is that the type is a union of data types for each page.

Is it possible to implement dynamic `ResultType` based on `QueryArgs`? Or is it a bad practice to merge all hooks into one and it's better to create a separate query for each endpoint?

**Comment by phryneas** [comment]
It's not possible. That's not really an RTKQ limitation, it's a TypeScript limitation: we need to map over those types to create new functions, and at that point is is just not possible to have any connection between `ArgType` and `ReturnType` .
